src/main/java/com/netdocuments/connect/kafka/handler/source/NDSourceHandler.java

```diff
@@ -109,99 +163,292 @@ public SourceRecordBuilder handle(SourceHandlerParams params) {
-      try {
-        newValue = JsonPropertyExtractor.extract(new ByteArrayInputStream(content),
-            fields.toArray(new String[fields.size()]));
-      } catch (Exception e) {
-        LOGGER.error("Error while extracting fields from document", e);
+      newValue = createMutationValue(docEvent, content);
+      if (newValue == null) {
         return false;
       }
-      if (!types.isEmpty()) {
-        if (newValue.containsKey("type") && !types.contains(newValue.get("type"))) {
-          LOGGER.info("type {} not in types {}", newValue.get("type"), types);
-          return false;
-        }
-      }
-      newValue.put("event", type.schemaName());
-      newValue.put("key", docEvent.key());
     } else {
       LOGGER.warn("unexpected event type {}", type);
       return false;
     }
 
     try {
-      byte[] value = objectMapper.writeValueAsBytes(newValue);
-      builder.value(null, value);
+      builder.value(null, convertToBytes(newValue, docEvent));
       return true;
-    } catch (JsonProcessingException e) {
+    } catch (DataException e) {
+      LOGGER.error("Failed to serialize data", e);
+      return false;
+    }
+  }
+
+  /**
+   * Creates a value map for deletion or expiration events.
+   */
+  private Map<String, Object> createDeletionOrExpirationValue(DocumentEvent docEvent, DocumentEvent.Type type) {
+    Map<String, Object> newValue = new HashMap<>();
+    newValue.put("event", type.schemaName());
+    newValue.put("key", docEvent.key());
+    return newValue;
+  }
+
+  /**
+   * Creates a value map for mutation events, extracting specified fields.
+   */
+  private Map<String, Object> createMutationValue(DocumentEvent docEvent, byte[] content) {
+    try {
+      Map<String, Object> newValue = JsonPropertyExtractor.extract(new ByteArrayInputStream(content),
+          fields.toArray(new String[fields.size()]));
+      newValue.put("event", DocumentEvent.Type.MUTATION.schemaName());
+      newValue.put("key", docEvent.key());
+      return newValue;
+    } catch (Exception e) {
+      LOGGER.error("Error while extracting fields from document", e);
+      return null;
+    }
+  }
+
+  /**
+   * Converts a value to bytes, applying CloudEvent format if necessary.
+   */
+  private byte[] convertToBytes(Map<String, Object> value, DocumentEvent docEvent) {
+    if (!cloudevent) {
+      return serializeToJson(value);
+    }
+    return withCloudEvent(serializeToJson(value), docEvent);
+  }
+
+  /**
+   * Serializes an object to JSON bytes.
+   */
+  private byte[] serializeToJson(Object value) {
+    try {
+      return OBJECT_MAPPER.writeValueAsBytes(value);
+    } catch (IOException e) {
       throw new DataException("Failed to serialize data", e);
     }
   }
+
+  /**
+   * Wraps the given value in a CloudEvent format.
+   */
+  private byte[] withCloudEvent(byte[] value, DocumentEvent documentEvent) {
+    Map<String, Object> cloudEventData = createCloudEventData(documentEvent);
```

Comment:
Or maybe this function could be done with a cloudevent library?

Reply:
Again it makes the code more complicated to use a library. You would think it would be straight forward but you have to do additional setup for the extra fields and you also have to figure out how to get it to serialize the way you want.